### PR TITLE
Add `read_text` and `read_blob` table functions

### PR DESCRIPTION
**Pull request by Maxxen**
This PR adds a set of new table functions, `read_text` and `read_blob`  that accepts a file name, list of file names or glob pattern, and outputs the content of each file as a varchar or blob, as well as additional metadata such as the file size and last modified time.

In `read_blob` the file content is returned as `BLOB`, but in `read_text` the file content is first validated to be valid UTF-8 and then returned as `VARCHAR`. If `read_text` attempts to read a file with invalid UTF-8 an error is thrown suggesting to use `read_blob` instead.

Example:
```sql
 SELECT size, parse_path(filename), content  FROM read_text('test/sql/table_function/files/*.txt');
┌───────┬───────────────────────────────────────────────┬──────────────┐
│ size  │             parse_path(filename)              │   content    │
│ int64 │                   varchar[]                   │   varchar    │
├───────┼───────────────────────────────────────────────┼──────────────┤
│    12 │ [test, sql, table_function, files, one.txt]   │ Hello World! │
│     2 │ [test, sql, table_function, files, three.txt] │ 42           │
│    10 │ [test, sql, table_function, files, two.txt]   │ Föö Bär      │
└───────┴───────────────────────────────────────────────┴──────────────┘

SELECT size, content, filename  FROM read_blob('test/sql/table_function/files/*');
┌───────┬──────────────────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────────┐
│ size  │                                             content                                              │                filename                 │
│ int64 │                                               blob                                               │                 varchar                 │
├───────┼──────────────────────────────────────────────────────────────────────────────────────────────────┼─────────────────────────────────────────┤
│   178 │ PK\x03\x04\x0A\x00\x00\x00\x00\x00\xACi=X\x14t\xCE\xC7\x0A\x00\x00\x00\x0A\x00\x00\x00\x09\x00…  │ test/sql/table_function/files/four.blob │
│    12 │ Hello World!                                                                                     │ test/sql/table_function/files/one.txt   │
│     2 │ 42                                                                                               │ test/sql/table_function/files/three.txt │
│    10 │ F\xC3\xB6\xC3\xB6 B\xC3\xA4r                                                                     │ test/sql/table_function/files/two.txt   │
└───────┴──────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────────┘


```

In cases where the underlying filesystem is unable to provide some of this data due to throwing a NotImplementedException (or a conversion error in the case of the `last_modified` column, because HTTPFS can't always return a valid timestamp), the cell is set to NULL instead.

This table function also utilizes projection pushdown so as to not compute properties unnecessarily. So you could e.g. use this to glob a directory full of huge files to get the file size in the `size` column, as long as you omit the `content` column the data wont be read into DuckDB.

### CODE REVIEW

**Comment by Tishj** [comment]
I feel for most purposes this would be used, VARCHAR is expected, maybe we can add an alternative `read_file_utf8` to support this?
Or even, since VARCHAR is likely the most commonly desired output, have `read_file` be VARCHAR and have `read_file_bytes` output BLOB?

As far as I know converting BLOB to VARCHAR is cumbersome.

**Comment by Maxxen** [comment]
> I feel for most purposes this would be used, VARCHAR is expected, maybe we can add an alternative `read_file_utf8` to support this? Or even, since VARCHAR is likely the most commonly desired output, have `read_file` be VARCHAR and have `read_file_bytes` output BLOB?
> 
> As far as I know converting BLOB to VARCHAR is cumbersome.

Hmm, I don't think adding a `decode` call is _that_ bad, but I am sympathetic to this. Maybe we could add a `read_text` table macro that performs the decode on the data column

**Comment by Tishj** [comment]
I forgot about `decode`, that's not too bad actually, but `read_text` sounds nice

**Review by carlopi — APPROVED** [review]
I tried to use it (and break it), looks to be working nicely, and last_modified can be handy.

Only concern I have is that `read_*` function names are becoming a scarce resource, and it would be cool to eventually have a read_XY(['path/to/fileA', 'path/to/fileB']) that automagically detect the common type and act on that.
But I think `read_file` is what this function is doing, so probably it's good. Plus there is a sqlite CLI functions that is called readfile and gets the blob out of a file, so it does match nicely.

**Comment by Maxxen** [comment]
I was gonna rewrite it read_blob, and then push an alias that tries to decode as text called read_text. Thoughts?

File: src/function/table/read_file.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+						FlatVector::SetNull(last_modified_vector, out_idx, true);
+					}
+				} break;
+				default:
+					FlatVector::SetNull(output.data[col_idx], out_idx, true);
+				}
+			}
+			// Filesystems are not required to support all operations, so we just set the column to NULL if not
+			// implemented
+			catch (NotImplementedException &) {
```
Comment (Mytherin) [inline src/function/table/read_file.cpp:179]: This should use the new `ErrorData` to correctly deal with RTTI in extensions in all cases

File: src/function/table/read_file.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	auto &bind_data = input.bind_data->Cast<ReadFileBindData>();
+	auto result = make_uniq<ReadFileGlobalState>();
+
+	result->files = bind_data.files;
+	result->current_file_idx = 0;
+	result->column_ids = input.column_ids;
+
+	for (const auto &column_id : input.column_ids) {
+		// For everything except the 'file' name column, we need to open the file
+		if (column_id > ReadFileBindData::FILE_NAME_COLUMN) {
```
Comment (Mytherin) [inline src/function/table/read_file.cpp:93]: `!=` is perhaps more clear

File: src/function/table/read_file.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+					FlatVector::GetData<int64_t>(file_size_vector)[out_idx] = file_handle->GetFileSize();
+				} break;
+				case ReadFileBindData::FILE_LAST_MODIFIED_COLUMN: {
+					auto &last_modified_vector = output.data[col_idx];
+					// This can sometimes fail (e.g. httpfs file system cant always parse the last modified time
+					// correctly)
+					try {
+						auto timestamp_seconds = Timestamp::FromEpochSeconds(fs.GetLastModifiedTime(*file_handle));
+						FlatVector::GetData<timestamp_t>(last_modified_vector)[out_idx] = timestamp_seconds;
+					} catch (ConversionException &) {
```
Comment (Mytherin) [inline src/function/table/read_file.cpp:169]: Same here - `ErrorData`

File: test/sql/table_function/read_text_and_blob.test [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+query III
+SELECT size, parse_path(filename), content  FROM read_text('test/sql/table_function/files/*.txt') ORDER BY filename, size;
+----
+12	[test, sql, table_function, files, one.txt]	Hello World!
+2	[test, sql, table_function, files, three.txt]	42
+10	[test, sql, table_function, files, two.txt]	Föö Bär
+
+# Test that the last_modified timestamp is reasonably correct
+query I
+SELECT last_modified > '2024-01-01' AND last_modified < '2500-01-01' FROM read_blob('test/sql/table_function/files/*');
```
Comment (Mytherin) [inline test/sql/table_function/read_text_and_blob.test:24]: Can we add some more test cases:

`SELECT COUNT(*) FROM read_text('test/sql/table_function/files/*.txt')`
`SELECT * FROM read_text('test/sql/table_function/files/nonexistentfile.txt')`
`SELECT filename FROM read_text('test/sql/table_function/files/nonexistentfile.txt')`
`SELECT filename FROM read_text(['test/sql/table_function/files/f1.txt', 'test/sql/table_function/files/f2.txt'])`

File: src/function/table/read_file.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+						FlatVector::SetNull(last_modified_vector, out_idx, true);
+					}
+				} break;
+				default:
+					FlatVector::SetNull(output.data[col_idx], out_idx, true);
```
Comment (Mytherin) [inline src/function/table/read_file.cpp:174]: `throw InternalException` is better here in the (likely) event we extend the function with more return columns

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some comments:

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**